iot-hub/Samples/device/MessageReceiveSample/MessageReceiveSample.cs

```diff
@@ -0,0 +1,108 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Diagnostics;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.Devices.Client.Samples
+{
+    public class MessageReceiveSample
+    {
+        private static readonly TimeSpan s_sleepDuration = TimeSpan.FromSeconds(5);
+        private static readonly TimeSpan s_receiveTimeout = TimeSpan.FromSeconds(10);
+        private readonly DeviceClient _deviceClient;
+        private readonly TransportType _transportType;
+
+        public MessageReceiveSample(DeviceClient deviceClient, TransportType transportType)
+        {
+            _deviceClient = deviceClient ?? throw new ArgumentNullException(nameof(deviceClient));
+            _transportType = transportType;
+        }
+
+        public async Task RunSampleAsync()
+        {
+            // First receive messages using the polling ReceiveAsync().
+            Console.WriteLine($"\n{DateTime.Now}> Device waiting for C2D messages from the hub for {s_receiveTimeout}...");
+            Console.WriteLine($"{DateTime.Now}> Use the Azure Portal IoT Hub blade or Azure IoT Explorer to send a message to this device.");
+            await ReceiveC2dMessagesPollingAndComplete(s_receiveTimeout);
+
+            // Now subscribe to receive C2D messages through a callback.
+            await _deviceClient.SetReceiveMessageHandlerAsync(OnC2dMessageReceived, _deviceClient);
+            Console.WriteLine($"\n{DateTime.Now}> Subscribed to receive C2D messages over callback.");
+
+            // Now wait to receive message from service.
+            // The message should be received on the callback, while the polling ReceiveAsync() call should return null.
+            Console.WriteLine($"\n{DateTime.Now}> Device waiting for C2D messages from the hub for {s_receiveTimeout}...");
+            Console.WriteLine($"{DateTime.Now}> Use the Azure Portal IoT Hub blade or Azure IoT Explorer to send a message to this device.");
+            await ReceiveC2dMessagesPollingAndComplete(s_receiveTimeout);
```

Comment:
Why polling if we're using a callback?

Reply:
I now get you are trying to illustrate that if a callback is enabled, the call to receive won't work. I really don't think there is value in demonstrating that in a sample.

Samples aren't for proving how a specific claim works, but showing customers how to write code to do something. We wouldn't want them to do this, so why demonstrate it?

Reply:
updated